### PR TITLE
Lru

### PR DESCRIPTION
**Pull request by Zagrevyum**
image

### CODE REVIEW

File: LRU.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+Follow up:
+Could you do both operations in O(1) time complexity?
+"""
+
+class LRUCache:
+
+    def __init__(self, capacity: int):
+        self.cache = {}
+        self.limit = capacity
```
Comment (TheKiqGit) [inline LRU.py]: `self.capacity` keep the same names otherwise it could be confusing

File: LRU.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    def update(self, key: int, value: int):
+        del [self.cache[key]]
```
Comment (TheKiqGit) [inline LRU.py]: redundant `[]`

File: LRU.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        self.cache[key] = value
+
+    def get(self, key: int) -> int:
+        if self.cache.get(key):
```
Comment (TheKiqGit) [inline LRU.py]: this works but you are wasting instructions, `get` gets you a value or a default (in this case None). so do this instead:

```
value = self.cache.get(key)  
if value:
    # do something
    return value
```

File: LRU.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            self.update(key, self.cache[key])
+            return self.cache[key]
+        else:
```
Comment (TheKiqGit) [inline LRU.py]: this works, but can be improved, I recommend avoiding conditions as much as possible, so your code is as flat as possible.


this is equivalent:
```
value = self.cache.get(key)  
if not value:
    return -1

# else
self.update(key, value)
return value
```

File: LRU.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            return -1
+
+    def put(self, key: int, value: int) -> None:
+        if not self.cache.get(key):
```
Comment (TheKiqGit) [inline LRU.py]: with the comments from above, see if you can streamline this

File: LRU.py (hunk 2 above)
Comment (TheKiqGit) [reply LRU.py]: you don't need to delete the key, you can just reassign the value

File: LRU.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+Could you do both operations in O(1) time complexity?
+"""
+
+class LRUCache:
+
+    def __init__(self, capacity: int):
+        self.cache = {}
+        self.capacity = capacity
+
+    def update(self, key: int, value: int):
```
Comment (TheKiqGit) [inline LRU.py:20]: you can remove this method, since you only need to do reassignment, you don't need to delete the current value

File: LRU.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            return value
+        return -1
+
+    def put(self, key: int, value: int) -> None:
+        if not self.cache.get(key):
+            if len(self.cache) < self.capacity:
+                self.cache[key] = value
+                return None
+            for k in self.cache:
+                del[self.cache[k]]
```
Comment (TheKiqGit) [inline LRU.py]: redundant `[]`

File: LRU.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+                break
+            self.cache[key] = value
+            return None
```
Comment (TheKiqGit) [inline LRU.py:40]: you can omit this if you re-structure your guard clause

File: LRU.py (hunk 2 above)
Comment (Zagrevyum) [reply LRU.py]: but that increases memory usage and breaks the logic for the capacity validation

Comment (Zagrevyum) [reply LRU.py]: or what do you mean by reassign?

File: LRU.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+    def get(self, key: int) -> int:
+        value = self.cache.get(key)
+        if value:
+            self.update(key, value)
+            return value
+        return -1
+
+    def put(self, key: int, value: int) -> None:
+        if not self.cache.get(key):
```
Comment (TheKiqGit) [inline LRU.py:32]: you could make this flatter if you do this instead:

```
curr_value = self.cache.get(key)
if curr_value:
    # we just need to update
    self.cache[key] = value
    return

if len(self.cache) < self.capacity:
    # we still have some room
    self.cache[key] = value
    return

# Eviction case
```

**Review by TheKiqGit — CHANGES_REQUESTED** [review]

File: LRU.py (hunk 2 above)
Comment (TheKiqGit) [reply LRU.py]: `self.cache[key] = new_value`

**Pull request merged**